File: memory_backends/file_backend.py
import json
from datetime import datetime, timezone
from pathlib import Path
from typing import Optional

from .interface import MemoryEntry, MemoryInterface


def _utcnow() -> str:
    return datetime.now(timezone.utc).isoformat()


class FileMemoryBackend(MemoryInterface):
    def __init__(self, base_path: str):
        self.base = Path(base_path)
        self.sessions_dir = self.base / "sessions"
        self.index_dir = self.base / "index"
        self.soul_path = self.base / "SOUL.md"
        self.user_path = self.base / "USER.md"
        self.entries_path = self.index_dir / "entries.jsonl"

    async def initialize(self) -> None:
        self.sessions_dir.mkdir(parents=True, exist_ok=True)
        self.index_dir.mkdir(parents=True, exist_ok=True)
        if not self.user_path.exists():
            self.user_path.write_text(
                "# User Profile\n\n## Identity\n\n## Preferences\n\n"
                "## Projects\n\n## Patterns\n\n## Context\n",
                encoding="utf-8",
            )
        if not self.entries_path.exists():
            self.entries_path.write_text("", encoding="utf-8")

    async def write(self, content: str, tags: Optional[list[str]] = None) -> None:
        tags = tags or []
        entry = {"content": content, "tags": tags, "timestamp": _utcnow()}
        with open(self.entries_path, "a", encoding="utf-8") as f:
            f.write(json.dumps(entry) + "\n")

    async def search(self, query: str, limit: int = 5) -> list[MemoryEntry]:
        if not self.entries_path.exists() or not self.entries_path.stat().st_size:
            return []
        query_words = set(query.lower().split())
        scored: list[tuple[int, str, dict]] = []  # (overlap, timestamp, data)
        with open(self.entries_path, encoding="utf-8") as f:
            for line in f:
                line = line.strip()
                if not line:
                    continue
                try:
                    data = json.loads(line)
                except json.JSONDecodeError:
                    continue
                content = data.get("content", "")
                if not content:
                    continue
                content_words = set(content.lower().split())
                overlap = len(query_words & content_words)
                if overlap > 0:
                    scored.append((overlap, data.get("timestamp", ""), data))
        scored.sort(key=lambda x: (x[0], x[1]), reverse=True)
        return [
            MemoryEntry(
                content=d["content"],
                tags=d.get("tags", []),
                timestamp=d.get("timestamp", ""),
                relevance=score / max(len(query_words), 1),
            )
            for score, _ts, d in scored[:limit]
        ]

    async def get_user_profile(self) -> str:
        if not self.user_path.exists():
            return ""
        return self.user_path.read_text(encoding="utf-8")

    async def update_user_profile(self, content: str) -> None:
        self.user_path.write_text(content, encoding="utf-8")

    async def get_soul(self) -> str:
        if not self.soul_path.exists():
            return "You are Atlas, a powerful personal AI system. Be direct, efficient, and proactive."
        return self.soul_path.read_text(encoding="utf-8")

    async def append_session(self, session_id: str, entry: dict) -> None:
        session_file = self.sessions_dir / f"{session_id}.jsonl"
        with open(session_file, "a", encoding="utf-8") as f:
            f.write(json.dumps(entry) + "\n")

    async def get_recent_sessions(self, n: int = 3) -> list[dict]:
        if not self.sessions_dir.exists():
            return []
        files = sorted(self.sessions_dir.glob("*.jsonl"), reverse=True)[:n]
        sessions = []
        for f in files:
            entries = []
            for line in f.read_text(encoding="utf-8").splitlines():
                line = line.strip()
                if line:
                    try:
                        entries.append(json.loads(line))
                    except json.JSONDecodeError:
                        pass
            if entries:
                sessions.append({"file": f.name, "entries": entries})
        return sessions

File: cli-ink/src/memory/file-backend.ts
import fs from "node:fs/promises";
import path from "node:path";
import { ATLAS_ROOT } from "../config/config.js";
import {
  DEFAULT_SOUL,
  DEFAULT_USER,
  type MemoryBackend,
  type MemoryHit,
  type Session,
  type SessionEntry,
} from "./backend.js";

export class FileMemoryBackend implements MemoryBackend {
  readonly kind = "file" as const;
  private readonly memoryDir: string;
  private readonly soulPath: string;
  private readonly userPath: string;
  private readonly sessionsDir: string;

  constructor(opts: { basePath?: string } = {}) {
    const base = opts.basePath ?? path.join(ATLAS_ROOT, "memory");
    this.memoryDir = base;
    this.soulPath = path.join(base, "SOUL.md");
    this.userPath = path.join(base, "USER.md");
    this.sessionsDir = path.join(base, "sessions");
  }

  async init(): Promise<void> {
    await fs.mkdir(this.memoryDir, { recursive: true });
    await fs.mkdir(this.sessionsDir, { recursive: true });
    await this.ensureFile(this.soulPath, DEFAULT_SOUL);
    await this.ensureFile(this.userPath, DEFAULT_USER);
  }

  private async ensureFile(p: string, fallback: string): Promise<string> {
    try {
      return await fs.readFile(p, "utf-8");
    } catch {
      await fs.mkdir(path.dirname(p), { recursive: true });
      await fs.writeFile(p, fallback);
      return fallback;
    }
  }

  async getSoul(): Promise<string> {
    return this.ensureFile(this.soulPath, DEFAULT_SOUL);
  }

  async getUserProfile(): Promise<string> {
    return this.ensureFile(this.userPath, DEFAULT_USER);
  }

  async updateUserProfile(content: string): Promise<void> {
    await fs.mkdir(this.memoryDir, { recursive: true });
    await fs.writeFile(this.userPath, content);
  }

  async appendSession(sessionId: string, entry: SessionEntry): Promise<void> {
    await fs.mkdir(this.sessionsDir, { recursive: true });
    const file = path.join(this.sessionsDir, `${sessionId}.jsonl`);
    await fs.appendFile(file, JSON.stringify(entry) + "\n");
  }

  async getRecentSessions(limit = 5): Promise<Session[]> {
    try {
      const files = await fs.readdir(this.sessionsDir);
      const jsonl = files.filter((f) => f.endsWith(".jsonl")).sort().reverse().slice(0, limit);
      const sessions: Session[] = [];
      for (const f of jsonl) {
        const s = await this.loadSession(f);
        if (s) sessions.push(s);
      }
      return sessions;
    } catch {
      return [];
    }
  }

  async loadSession(file: string): Promise<Session | null> {
    try {
      const text = await fs.readFile(path.join(this.sessionsDir, file), "utf-8");
      const entries = text
        .split("\n")
        .filter(Boolean)
        .map((l) => {
          try {
            return JSON.parse(l) as SessionEntry;
          } catch {
            return null;
          }
        })
        .filter((e): e is SessionEntry => e !== null);
      return { file, entries };
    } catch {
      return null;
    }
  }

  async search(query: string, limit = 5): Promise<MemoryHit[]> {
    const sessions = await this.getRecentSessions(20);
    const queryTerms = new Set(query.toLowerCase().split(/\s+/).filter(Boolean));
    if (queryTerms.size === 0) return [];

    const scored: { hit: MemoryHit; score: number }[] = [];
    for (const s of sessions) {
      for (const e of s.entries) {
        const text = `${e.user} ${e.assistant}`.toLowerCase();
        let score = 0;
        for (const t of queryTerms) if (text.includes(t)) score += 1;
        if (score > 0) {
          scored.push({
            hit: {
              content: `User: ${e.user.slice(0, 200)} / Atlas: ${e.assistant.slice(0, 200)}`,
              timestamp: e.timestamp,
              relevance: score / queryTerms.size,
            },
            score,
          });
        }
      }
    }
    scored.sort((a, b) => b.score - a.score);
    return scored.slice(0, limit).map((s) => s.hit);
  }
}

// Re-export the types for callers that imported from here historically.
export type { Session, SessionEntry, MemoryHit } from "./backend.js";
